sdk/core/azure-core/src/base64.cpp

```diff
@@ -4,93 +4,474 @@
+
+  int32_t i0 = Base64EncodeArray[i >> 18];
+  int32_t i1 = Base64EncodeArray[(i >> 12) & 0x3F];
+  int32_t i2 = Base64EncodeArray[(i >> 6) & 0x3F];
+  int32_t i3 = Base64EncodeArray[i & 0x3F];
+
+  return i0 | (i1 << 8) | (i2 << 16) | (i3 << 24);
+}
+
+static int32_t base64EncodeAndPadOne(const uint8_t* twoBytes)
+{
+  int32_t i = (*twoBytes << 16) | (*(twoBytes + 1) << 8);
+
+  int32_t i0 = Base64EncodeArray[i >> 18];
+  int32_t i1 = Base64EncodeArray[(i >> 12) & 0x3F];
+  int32_t i2 = Base64EncodeArray[(i >> 6) & 0x3F];
+
+  return i0 | (i1 << 8) | (i2 << 16) | (EncodingPath << 24);
+}
+
+static int32_t base64EncodeAndPadTwo(const uint8_t* oneByte)
+{
+  int32_t i = (*oneByte << 8);
+
+  int32_t i0 = Base64EncodeArray[i >> 10];
+  int32_t i1 = Base64EncodeArray[(i >> 4) & 0x3F];
+
+  return i0 | (i1 << 8) | (EncodingPath << 16) | (EncodingPath << 24);
+}
+
+static void base64WriteIntAsFourBytes(char* destination, int32_t value)
+{
+  *(destination + 3) = static_cast<uint8_t>((value >> 24) & 0xFF);
+  *(destination + 2) = static_cast<uint8_t>((value >> 16) & 0xFF);
+  *(destination + 1) = static_cast<uint8_t>((value >> 8) & 0xFF);
+  *(destination + 0) = static_cast<uint8_t>(value & 0xFF);
+}
+
+std::string base64Encode(const std::vector<uint8_t>& data)
+{
+  int64_t sourceIndex = 0;
+  auto inputSize = data.size();
+  int32_t result = 0;
+  auto maxEncodedSize = ((inputSize + 2) / 3) * 4;
+  // Use a string with size to the max possible result
+  std::string encodedResult(maxEncodedSize, '0');
+  // Removing const from the string to update the placeholder string
+  auto destination = const_cast<char*>(encodedResult.data());
+
+  while (sourceIndex < static_cast<int64_t>(inputSize) - 2)
+  {
+    result = base64Encode(data.data() + sourceIndex);
+    base64WriteIntAsFourBytes(destination, result);
+    destination += 4;
+    sourceIndex += 3;
+  }
+
+  if (sourceIndex == static_cast<int64_t>(inputSize) - 1)
+  {
+    result = base64EncodeAndPadTwo(data.data() + sourceIndex);
+    base64WriteIntAsFourBytes(destination, result);
+    destination += 4;
+    sourceIndex += 1;
+  }
+  else if (sourceIndex == static_cast<int64_t>(inputSize) - 2)
+  {
+    result = base64EncodeAndPadOne(data.data() + sourceIndex);
+    base64WriteIntAsFourBytes(destination, result);
+    destination += 4;
+    sourceIndex += 2;
+  }
+  auto destinationLength = static_cast<size_t>(destination - encodedResult.data());
+  // If encoding took less than the max-expected
+  if (destinationLength < encodedResult.size())
+  {
+    return encodedResult.substr(0, destinationLength);
+  }
+  return encodedResult;
+}
 
-#if defined(AZ_PLATFORM_WINDOWS)
+static int64_t base64Decode(const char* encodedBytes)
+{
+  int64_t i0 = *encodedBytes;
+  int64_t i1 = *(encodedBytes + 1);
+  int64_t i2 = *(encodedBytes + 2);
+  int64_t i3 = *(encodedBytes + 3);
 
-  std::string Convert::Base64Encode(const std::vector<uint8_t>& data)
+  i0 = Base64DecodeArray[i0];
```

Comment:
On platforms where `char` is signed (C and C++ let each implementation decide if `char` is signed or not) this could be accessing memory _before_ `Base64DecodeArray`. I recommend:
```c++
int32_t i0 = Base64DecodeArray[static_cast<unsigned char>(encodedBytes[0])];
int32_t i1 = Base64DecodeArray[static_cast<unsigned char>(encodedBytes[1])];
int32_t i2 = Base64DecodeArray[static_cast<unsigned char>(encodedBytes[2])];
int32_t i3 = Base64DecodeArray[static_cast<unsigned char>(encodedBytes[3])];
int32_t errcheck = (i0 | i1 | i2 | i3) & 0xC0); // If either of the high two bits are set in any result, the input is invalid
return (errcheck << 24) | i0 << 18 | i1 << 12 | i2 << 6 | i3; // If any of the high 8 bits is set, the result is invalid
```

Reply:
> C and C++ let each implementation decide if `char` is signed or not

🤦‍♂️

Why?! Lol

Reply:
Pre-standardization implementation divergence, to my understanding.

Reply:
Good thing the C impl is using `int32_t` then instead of char:
https://github.com/Azure/azure-sdk-for-c/blob/7f689bea833a12b0be9059590e7321d5af99296f/sdk/src/azure/core/az_base64.c#L380-L383

Though, now I wonder what invalid assumptions I made about C types when writing that code and whether that code could also benefit from another set of eyes to review :)